src/inet/InetInterface.cpp

```diff
@@ -119,235 +154,760 @@ NL_DLL_EXPORT INET_ERROR InterfaceNameToId(const char *intfName, InterfaceId& in
 
-    if (rv != -1) {
-        curAddr = addrsList;
-    } else {
-        curAddr = addrsList = NULL;
+InterfaceIterator::InterfaceIterator(void)
+{
+    mIntfArray = NULL;
+    mCurIntf = 0;
+    mIntfFlags = 0;
+    mIntfFlagsCached = 0;
+}
+
+#endif // WEAVE_SYSTEM_CONFIG_USE_SOCKETS
+
+/**
+ * @fn      InterfaceIterator::~InterfaceIterator(void)
+ *
+ * @brief   Destroys an InterfaceIterator object.
+ *
+ * @details
+ *     Recycles any resources allocated by the constructor.
+ */
+
+#if WEAVE_SYSTEM_CONFIG_USE_SOCKETS
+
+InterfaceIterator::~InterfaceIterator(void)
+{
+    if (mIntfArray != NULL)
+    {
+        if_freenameindex(mIntfArray);
+        mIntfArray = NULL;
     }
+}
+
 #endif // WEAVE_SYSTEM_CONFIG_USE_SOCKETS
+
+/**
+ * @fn      bool InterfaceIterator::HasCurrent(void)
+ *
+ * @brief   Test whether the iterator is positioned on an interface
+ *
+ * @return  \c true if the iterator is positioned on an interface;
+ *          \c false if positioned beyond the end of the interface list.
+ */
+
+#if WEAVE_SYSTEM_CONFIG_USE_SOCKETS
+
+bool InterfaceIterator::HasCurrent(void)
+{
+    return (mIntfArray != NULL)
+        ? mIntfArray[mCurIntf].if_index != 0
+        : Next();
 }
 
-InterfaceIteratorBasis::~InterfaceIteratorBasis(void)
+#endif // WEAVE_SYSTEM_CONFIG_USE_SOCKETS
+
+/**
+ * @fn      bool InterfaceIterator::Next(void)
+ *
+ * @brief   Advance the iterator to the next network interface.
+ *
+ * @return  \c false if advanced beyond the end, else \c true.
+ *
+ * @details
+ *     Advances the internal iterator to the next network interface or to a position
+ *     beyond the end of the interface list.
+ *
+ *     On multi-threaded LwIP systems, this method is thread-safe relative to other
+ *     threads accessing the global LwIP state provided that: 1) the other threads
+ *     hold the LwIP core lock while mutating the list of netifs; and 2) netif objects
+ *     themselves are never destroyed.
+ *
+ *     Iteration is stable in the face of changes to the underlying system's
+ *     interfaces, *except* in the case of LwIP systems when the currently selected
+ *     interface is removed from the list, which causes iteration to end immediately.
+ */
+bool InterfaceIterator::Next(void)
 {
+#if WEAVE_SYSTEM_CONFIG_USE_SOCKETS
+
+    if (mIntfArray == NULL)
+    {
+        mIntfArray = if_nameindex();
+    }
+    else if (mIntfArray[mCurIntf].if_index != 0)
+    {
+        mCurIntf++;
+        mIntfFlags = 0;
+        mIntfFlagsCached = false;
+    }
+    return (mIntfArray != NULL && mIntfArray[mCurIntf].if_index != 0);
+
+#endif // WEAVE_SYSTEM_CONFIG_USE_SOCKETS
+
 #if WEAVE_SYSTEM_CONFIG_USE_LWIP
-    curIntf = NULL;
+
+    // Lock LwIP stack
+    LOCK_TCPIP_CORE();
+
+    // Verify the previous netif is still on the list if netifs.  If so,
+    // advance to the next nextif.
+    struct netif * prevNetif = mCurNetif;
+    for (mCurNetif = netif_list; mCurNetif != NULL; mCurNetif = mCurNetif->next)
```

Comment:
IIRC, using the newer NETIF_FOREACH, when available, as demonstrated in IPEndPoint.h allows for an easier single-netif optimization in LwIP.

Reply:
Thanks.  I wasn't aware of that pattern.  Code updated accordingly.